pkg/kubelet/logs/container_log_manager.go

```diff
@@ -213,63 +223,93 @@ func (c *containerLogManager) Clean(ctx context.Context, containerID string) err
 	return nil
 }
 
+func (c *containerLogManager) processQueueItems(worker int) {
+	klog.V(4).InfoS("Starting container log rotate worker", "workerID", worker)
+	for c.processContainer(context.Background(), worker) {
+	}
+	klog.V(4).InfoS("Terminating container log rotate worker", "workerID", worker)
+}
+
 func (c *containerLogManager) rotateLogs(ctx context.Context) error {
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
+	klog.V(4).InfoS("Starting container log rotation sequence")
 	// TODO(#59998): Use kubelet pod cache.
 	containers, err := c.runtimeService.ListContainers(ctx, &runtimeapi.ContainerFilter{})
 	if err != nil {
 		return fmt.Errorf("failed to list containers: %v", err)
 	}
-	// NOTE(random-liu): Figure out whether we need to rotate container logs in parallel.
 	for _, container := range containers {
 		// Only rotate logs for running containers. Non-running containers won't
 		// generate new output, it doesn't make sense to keep an empty latest log.
 		if container.GetState() != runtimeapi.ContainerState_CONTAINER_RUNNING {
 			continue
 		}
-		id := container.GetId()
-		// Note that we should not block log rotate for an error of a single container.
-		resp, err := c.runtimeService.ContainerStatus(ctx, id, false)
-		if err != nil {
-			klog.ErrorS(err, "Failed to get container status", "containerID", id)
-			continue
-		}
-		if resp.GetStatus() == nil {
-			klog.ErrorS(err, "Container status is nil", "containerID", id)
-			continue
+		// Doing this to avoid additional overhead with logging of label like arguments that can prove costly
+		if v := klog.V(4); v.Enabled() {
+			klog.V(4).InfoS("Adding new entry to the queue for processing", "id", container.GetId(), "name", container.Metadata.GetName(), "labels", container.GetLabels())
 		}
-		path := resp.GetStatus().GetLogPath()
-		info, err := c.osInterface.Stat(path)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				klog.ErrorS(err, "Failed to stat container log", "path", path)
-				continue
-			}
-			// In rotateLatestLog, there are several cases that we may
-			// lose original container log after ReopenContainerLog fails.
-			// We try to recover it by reopening container log.
-			if err := c.runtimeService.ReopenContainerLog(ctx, id); err != nil {
-				klog.ErrorS(err, "Container log doesn't exist, reopen container log failed", "containerID", id, "path", path)
-				continue
-			}
-			// The container log should be recovered.
-			info, err = c.osInterface.Stat(path)
-			if err != nil {
-				klog.ErrorS(err, "Failed to stat container log after reopen", "path", path)
-				continue
-			}
+		c.queue.Add(container.GetId())
+	}
+	wait.PollInfinite(20*time.Millisecond, func() (done bool, err error) {
+		return c.queue.Len() == 0, nil
```

Comment:
do we have to wait until `== 0`? how about `<= ContainerLogMaxWorkers`? (if we have more workers than what's in the queue?)

i am also thinking about the processing order, if we have a lot of containers, does it matter where in the list order they are? Do we need to prioritize containers that spew a lot more than others? also how long can this queue get? if the queue is long we will still end up starving log gathering from containers that are putting things out fast no?

(Sorry just thinking aloud!)

Reply:
> do we have to wait until == 0? how about <= ContainerLogMaxWorkers? 

The idea of waiting for 0 was to avoid any possible race condition that might arise by running rotate on the same container log from two different worker. without having to track the ongoing rotation, I couldn't think of a simple way to achieve that. But I am definitely open to suggestion on what might be a better way to do this without having to wait for the queue to shrink to 0 size before picking the next batch. 

> does it matter where in the list order they are

Yes and no. It matters slightly in terms of time taken for rotating such container log over others. However, for anything else, the ordering doesn't seem to matter. If you are thinking of any other cases where it might matter, we can discuss that.

> Do we need to prioritize containers that spew a lot more than others?

I think, from inside this logic we might not be able to do that effectively. If ever we want to do something like that, I would prefer a processing priority assignment to the pod by means of an annotation may be ? That way, end users can control this based on their needs. (I might be over thinking this as well. So please feel free to correct me if you think there is a better way for this)

> if the queue is long we will still end up starving log gathering from containers that are putting things out fast no?

Yes, You are absolutely right on that matter. In such cases one would have to tune the worker count to be much higher to account for such needs. I believe this will be a function of number of pods per node vs the worker count. 

Let us see, 

You have a duration of 5 second configured, with 10 worker and your cluster has a per node pod density of 100.

Queue will be refurbished every 5 second provided that the queue is empty at the end of the 5th second. 
On a best case, each 10 worker, taking uniform time to rotate logs, will process 10 pod logs per iteration of the queue population interval.

Now, when some worker get loaded with pods that generated too much logs, this will invariably increase the time taken to process such item. When that happens, your queue drain to 0 takes much higher than the configured duration of re-population. 

So, this, now leaves a window of opportunity for the pods that generate too much logs to overshoot their allowed limit as the next rotation will happen only once the queue is re-populated. 

Another thing to keep in mind is, even if the pod is pushed into the processing queue, it might still take a bit before it is picked up and rotated. So the window of it getting picked up, put into queue, get picked from the queue and processing is done, there will be a window for it to generate enough logs to overshoot the allowed limit. 

Thanks really valid questions @dims  I have answered them the best I can think of, But I would prefer a second opinion from you and others in sig-node.

Reply:
thanks @harshanarayana